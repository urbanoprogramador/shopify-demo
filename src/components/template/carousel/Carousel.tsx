import React from 'react';
import { useState, useEffect, useContext, createContext } from 'react';
import { useSelector } from 'react-redux';
import { selectThemeconfig } from '../../../core/store/theme/selectors/selector';



//interfas context

interface ICarousel{
        active:any,
        setActive:any,
        items:any,
        isImg:any,
}
export const CarrouselContext = createContext({} as ICarousel);


const Item = ({
    item
}) => {

    const { items, isImg } = useContext(CarrouselContext);

    return (<div className='itemCarrusel ' style={{ width: (100 / items.length) + '%' }}>
        {
            isImg ? <img src={item} alt='imagen del producto' /> : item
        }
    </div>);
}

const FooterCarousel = () => {
    //720px width
    const widthScreen = useSelector(selectThemeconfig).width;

    const { active, items, setActive, isImg } = useContext(CarrouselContext);

    const [points, setPoints] = useState<number[]>([]);

    useEffect(() => {

        let array:number[] = [];

        if (widthScreen <= 768) {
            for (let i = active - 2; i < active + 3; i++) {
                let index = i;
                if (i < 0) {
                    index = items.length + i;
                } else if (i >= items.length) {
                    index = i % items.length;
                }
                array.push(index);
            }
        } else {
            for (let i = active - 1; i < active + 2; i++) {
                let index = i;
                if (i < 0) {
                    index = items.length + i;
                } else if (i >= items.length) {
                    index = i % items.length;
                }
                array.push(index);
            }
        }

        setPoints(array);
    }, [active, setPoints, widthScreen, items]);



    return (<>
        <div className='bodyMinuatura'>
            {

                points.map((e, indice) => {
                    return <div onClick={() => setActive(e)} className='bodyMinuaturaItem' key={indice}>

                        {
                            isImg ?
                                <img src={items[e]} alt='miniatura del producto' />
                                : items[e]
                        }
                    </div>;
                })
            }

        </div>

    </>);
}


export const Carousel = ({ items, isImg }:{items:any[],isImg:boolean}) => {

    const [active, setActive] = useState<number>(0);
    console.log('estoy entrando ');
    
    return (<CarrouselContext.Provider value={{
        active,
        setActive,
        items,
        isImg
    }}>

        <div className="BodyCarrusel">
            <div className='ContenCarrousel' style={{ width: (100 * items.length) + "%", transform: "translateX(-" + ((100 / items.length) * active) + "%)" }}>
                {
                    items.map((e, indice) => {
                        return <Item key={indice} item={e} />
                    })
                }
            </div>
            <FooterCarousel />
        </div>
    </CarrouselContext.Provider>);
}