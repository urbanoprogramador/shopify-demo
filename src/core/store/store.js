import { middlewareReduxInpect} from './middleware/middleware';
import { combineReducers, createStore } from 'redux';
import { userReducer } from './auth/reducers/reducer';
import { productReducer } from './product/reducers/reducer';
import { themeReducer } from './theme/reducers/reducer';


const reducer=combineReducers({
  auth:userReducer,
  product:productReducer,
  theme:themeReducer
})

export const store = createStore(reducer,middlewareReduxInpect);