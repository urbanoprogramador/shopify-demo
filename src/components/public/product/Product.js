import { createContext, useContext, useEffect, useState } from "react";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from 'react-redux';
import { asyncActionLoadProduct } from "../../../core/store/product/actions/action";
import { selectGalery, selectOptions, selectPrice, selectProductStatus, selectTitle, selectVendor, selectDescription } from './../../../core/store/product/selectors/selector';
import { Carousel } from "../../template/carousel/Carousel";
import { clearHtml } from "../../../core/helper/clearHtml";
import { Modal } from "../../portal/Modal";


const formatter = new Intl.NumberFormat('en-US', {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
});


export const CartContext = createContext(null);

const Galeria = () => {

    const galery = useSelector(selectGalery);

    return (<>
        <Carousel isImg={true} items={galery} />
    </>);
}

const Seccion1 = () => {
    const vendor = useSelector(selectVendor);
    const title = useSelector(selectTitle);
    const price = useSelector(selectPrice);

    return (<div className="headerProduct">
        <h3>{vendor}</h3>
        <h1>{title}</h1>
        <h2>${formatter.format(price.price_min / 100)} <span>${formatter.format(price.compare_at_price / 100)}</span></h2>
    </div>);
}

const SeccionColor = () => {
    const color = useSelector(selectOptions('Color'));

    const { color: selecter, setColor, } = useContext(CartContext);


    return (<div className="contectOpction">
        <div className="rowLabel">
            Color:
        </div>
        <div className="rowValue">
            {
                color.values.map((e, index) => {
                    return (<span key={e} className={`selectColorBorder ${selecter === index ? 'active' : ''}`} onClick={() => { setColor(index) }}>
                        <span key={e} style={{ background: e }} className="selectColor"></span>
                    </span>);
                })
            }
        </div>
    </div>);
}

const SessionSize = () => {
    const size = useSelector(selectOptions('Size'));

    const { size: selecter, setSize, } = useContext(CartContext);
    return (<div className="contectOpction">
        <div className="rowLabel">
            Size:
        </div>
        <div className="rowValue">
            {
                size.values.map((e, index) => {
                    return (< span key={e} className={`SelectSize ${selecter === index ? 'active' : ''}`} onClick={() => { setSize(index) }}>
                        {e}
                    </span>);
                })
            }
        </div>
    </div>);
}

const SeccionCantidad = () => {
    const price = useSelector(selectPrice);

    const { cantidad,
        setCantidad } = useContext(CartContext);


    return (
        <div className="justify2">

            <div className="buttonNumero">
                <button onClick={() => { setCantidad(e => e - 1 < 0 ? 0 : e - 1) }}>-</button>
                <input type="text" value={cantidad} onChange={({ target }) => {
                    console.log(target.value);
                    setCantidad(parseInt(target.value));
                }} />
                <button onClick={() => { setCantidad(e => e + 1) }}>+</button>
            </div>
            <div className="rigth">
                total $ {(price.price_min / 100) * cantidad}
            </div>
        </div>);
}


const AddCartItem = ({ close }) => {
    const { size: sizeIndex, color: colorIndex, cantidad } = useContext(CartContext);
    const color = useSelector(selectOptions('Color'));
    const size = useSelector(selectOptions('Size'));

    const price = useSelector(selectPrice);

    const galery = useSelector(selectGalery);


    console.log({
        color, size
    });


    return (<Modal>

        <div className="contentModal">
            <div className="titelModal">
                <h1>You want to add this item to the shopping cart? <img style={{width:'70px'}} src={galery[0]} alt='imagen del producto' /></h1>
            </div>
            <div className="bodyModal">
                <div>
                    Color: {color.values[colorIndex]}
                </div>
                <div>
                    Size: {size.values[sizeIndex]}
                </div>
                <h4>
                Quantity : {cantidad} Total {price.price_min*cantidad}
                </h4>
            </div>

            <div className="footerModal">
                <div className="bottonera">

                <button className="btn" onClick={() => { close(false) }}>Cancel </button>
                <button className="btn black" onClick={() => { close(false) }}>Accept</button>
                </div>
            </div>

        </div>
    </Modal>);
}

const ButtonAdd = () => {
    const description = useSelector(selectDescription);

    const [isModalActive, setIsModalActive] = useState(false);

    const { cantidad } = useContext(CartContext);

    const handleAddCart = () => {
        setIsModalActive(true);
    }

    useEffect(() => {
        let time;
        if (isModalActive) {
            time = setTimeout(() => {
                setIsModalActive(false);
            }, 300000);
        }
        return () => {
            clearTimeout(time);
        }
    }, [isModalActive, setIsModalActive]);


    return (
        <div className="">

            <div >
                <button className="btn btn-2" >Add to Favorite</button>
                <button className="btn btn-2 black" onClick={handleAddCart} disabled={cantidad<1} >Add to Cart</button>
            </div>
            <p>
                {clearHtml(description)}
            </p>
            {
                isModalActive ? <AddCartItem close={setIsModalActive}></AddCartItem> : null
            }

        </div>);
}

const Detalle = () => {
    return (<>
        <Seccion1 />
        <hr />
        <SeccionColor />
        <hr />
        <SessionSize />
        <hr />
        <SeccionCantidad />
        <ButtonAdd />
    </>);
}



export const Product = () => {

    const { product } = useParams();
    const [unica, setUnica] = useState(true);
    const dispatch = useDispatch();
    const status = useSelector(selectProductStatus);


    const [color, setColor] = useState(0);
    const [size, setSize] = useState(0);
    const [cantidad, setCantidad] = useState(0);

    useEffect(() => {
        const stop = setTimeout(() => {
            dispatch(asyncActionLoadProduct(`https://graditest-store.myshopify.com/products/${product}.js`));
        }, 1500);
        return () => {
            clearTimeout(stop);
        }
    }, [product, setUnica, unica, dispatch]);


    switch (status.loading) {
        case 'succeded':
            return (<CartContext.Provider value={{
                color,
                setColor,
                size,
                setSize,
                cantidad,
                setCantidad,
            }} >

                <div className="contentProduct">
                    <div className="row">
                        <Galeria />
                    </div>
                    <div className="row">
                        <Detalle />
                    </div>
                </div>
            </CartContext.Provider>);
        case 'rejected':
            return (<>

                <h2>Error </h2>
            </>);
        default:
            return (<>

<Modal>

<div className="contentModal">
    <div className="titelModal">
        <h1 style={{textAlaing:'center'}}>Loading...</h1>
    </div>




</div>
</Modal>
            </>);
    }


}