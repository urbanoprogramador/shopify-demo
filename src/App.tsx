import React from 'react';
import { Provider } from 'react-redux';
import './core/styles/styles.scss';
import { store } from './core/store/store';
import { AppRouter } from './routers/AppRouter';
import { actionLoadConfigTheme } from './core/store/theme/actions/action';


start();
window.addEventListener('resize', start);

function start(){
  store.dispatch(actionLoadConfigTheme({width:document.documentElement.clientWidth}));
}


function App():JSX.Element {
  return (
    <Provider store={store}>
      <AppRouter />
    </Provider>
  );
}

export default App;
