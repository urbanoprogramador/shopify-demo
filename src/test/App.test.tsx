import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
jest.setTimeout(10000);
import App from '../App';

import '@testing-library/jest-dom'





/*

import { rest } from 'msw'
import { setupServer } from 'msw/node'

export const handlers = [
  rest.get('/api/user', (req, res, ctx) => {
    return res(ctx.json('John Smith'), ctx.delay(150))
  })
]

const server = setupServer(...handlers); */

import { Provider } from 'react-redux';
import { Product } from 'components/public/product/Product';
import { store } from 'core/store/store';
import { Carousel } from 'components/template/carousel/Carousel';
import { MemoryRouter, Route } from 'react-router';
import { Routes } from 'react-router-dom';


describe("Probando el enrutamiento", () => {


  /*   beforeAll(() => server.listen())
    afterEach(() => server.resetHandlers())
    afterAll(() => server.close()) */



  test('Prueba uno ', () => {
    const mensaje = 'Holamundo';

    expect(mensaje).toMatchSnapshot();


  });


  test('Prueba 2', () => {
    render(<App />);

    expect(screen.debug()).toMatchSnapshot();
  });


  test('prueba de product', async () => {

    render(<Provider store={store}>
      <MemoryRouter initialEntries={['/product/free-trainer-3-mmw']}>
        <Routes>
          <Route path='/product/:product' element={<Product />} />

        </Routes>
      </MemoryRouter>
    </Provider>
    );
    /*     await waitFor(() => expect(screen.getByTestId('producto')).toMatchSnapshot(), {
          timeout: 8000,
    
        }); */
    await waitFor(() => expect(screen.getByTestId('producto')).toMatchSnapshot(), {
      timeout: 8000,

    });
  });
  test('prueba de Carousel', async () => {
    render(
      <Provider store={store}>
        <Carousel isImg={true} items={['//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-4.jpg?v=1639585159',
          '//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-3.jpg?v=1639585158',
          '//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-2.jpg?v=1639585158',
          '//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-1.jpg?v=1639585159']} />
      </Provider>
    );
    expect(screen.getByTestId('carrusel')).toMatchSnapshot()
    /* await waitFor(() => expect(screen.getByTestId('producto')).toMatchSnapshot(), {
      timeout: 8000,

    }); */
  });

  test('probando al clickear en cantidad de productos', async () => {
    render(<Provider store={store}>
      <MemoryRouter initialEntries={['/product/free-trainer-3-mmw']}>
        <Routes>
          <Route path='/product/:product' element={<Product />} />

        </Routes>
      </MemoryRouter>
    </Provider>
    );
    /*     await waitFor(() => expect(screen.getByTestId('producto')).toMatchSnapshot(), {
          timeout: 8000,
    
        }); */
    await waitFor(() => {}, {
      timeout: 8000,
    });
    const el=document.querySelector('.addCartTesting');
    if(el){
      fireEvent.click(el);
      const input=document.getElementsByClassName('CantidadTesting')[0] as HTMLInputElement;
      expect(input.value ).toBe('1');

    }else{
      throw 'deveria existir el boton';
    }
  });







});
