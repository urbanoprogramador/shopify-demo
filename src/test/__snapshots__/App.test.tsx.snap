// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el enrutamiento Prueba 2 1`] = `undefined`;

exports[`Probando el enrutamiento Prueba uno  1`] = `"Holamundo"`;

exports[`Probando el enrutamiento prueba de Carousel 1`] = `
<div
  class="BodyCarrusel"
  data-testid="carrusel"
>
  <div
    class="ContenCarrousel"
    style="width: 400%; transform: translateX(-0%);"
  >
    <div
      class="itemCarrusel "
      style="width: 25%;"
    >
      <img
        alt="imagen del producto"
        src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-4.jpg?v=1639585159"
      />
    </div>
    <div
      class="itemCarrusel "
      style="width: 25%;"
    >
      <img
        alt="imagen del producto"
        src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-3.jpg?v=1639585158"
      />
    </div>
    <div
      class="itemCarrusel "
      style="width: 25%;"
    >
      <img
        alt="imagen del producto"
        src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-2.jpg?v=1639585158"
      />
    </div>
    <div
      class="itemCarrusel "
      style="width: 25%;"
    >
      <img
        alt="imagen del producto"
        src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-1.jpg?v=1639585159"
      />
    </div>
  </div>
  <div
    class="bodyMinuatura"
  >
    <div
      class="bodyMinuaturaItem"
    >
      <img
        alt="miniatura del producto"
        src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-1.jpg?v=1639585159"
      />
    </div>
    <div
      class="bodyMinuaturaItem"
    >
      <img
        alt="miniatura del producto"
        src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-4.jpg?v=1639585159"
      />
    </div>
    <div
      class="bodyMinuaturaItem"
    >
      <img
        alt="miniatura del producto"
        src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-3.jpg?v=1639585158"
      />
    </div>
  </div>
</div>
`;

exports[`Probando el enrutamiento prueba de product 1`] = `
<div
  class="contentProduct"
  data-testid="producto"
  id="demo"
>
  <div
    class="row"
  >
    <div
      class="BodyCarrusel"
      data-testid="carrusel"
    >
      <div
        class="ContenCarrousel"
        style="width: 400%; transform: translateX(-0%);"
      >
        <div
          class="itemCarrusel "
          style="width: 25%;"
        >
          <img
            alt="imagen del producto"
            src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-1.jpg?v=1639585159"
          />
        </div>
        <div
          class="itemCarrusel "
          style="width: 25%;"
        >
          <img
            alt="imagen del producto"
            src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-2.jpg?v=1639585158"
          />
        </div>
        <div
          class="itemCarrusel "
          style="width: 25%;"
        >
          <img
            alt="imagen del producto"
            src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-3.jpg?v=1639585158"
          />
        </div>
        <div
          class="itemCarrusel "
          style="width: 25%;"
        >
          <img
            alt="imagen del producto"
            src="//cdn.shopify.com/s/files/1/0454/8316/3809/products/product-4.jpg?v=1639585159"
          />
        </div>
      </div>
      <div
        class="bodyMinuatura"
      />
    </div>
  </div>
  <div
    class="row"
  >
    <div
      class="headerProduct"
    >
      <h3>
        Graditest Store
      </h3>
      <h1>
        Free Trainer 3 / MMW
      </h1>
      <h2>
        $
        285.00
         
        <span>
          $
          370.00
        </span>
      </h2>
    </div>
    <hr />
    <div
      class="contectOpction"
    >
      <div
        class="rowLabel"
      >
        Color:
      </div>
      <div
        class="rowValue"
      >
        <span
          class="selectColorBorder active"
        >
          <span
            class="selectColor"
            style="background: Red;"
          />
        </span>
        <span
          class="selectColorBorder "
        >
          <span
            class="selectColor"
            style="background: Black;"
          />
        </span>
      </div>
    </div>
    <hr />
    <div
      class="contectOpction"
    >
      <div
        class="rowLabel"
      >
        Size:
      </div>
      <div
        class="rowValue"
      >
        <span
          class="SelectSize active"
        >
          7
        </span>
        <span
          class="SelectSize "
        >
          7.5
        </span>
        <span
          class="SelectSize "
        >
          8
        </span>
        <span
          class="SelectSize "
        >
          8.5
        </span>
        <span
          class="SelectSize "
        >
          9
        </span>
        <span
          class="SelectSize "
        >
          9.5
        </span>
        <span
          class="SelectSize "
        >
          10
        </span>
        <span
          class="SelectSize "
        >
          10.5
        </span>
        <span
          class="SelectSize "
        >
          11
        </span>
        <span
          class="SelectSize "
        >
          11.5
        </span>
      </div>
    </div>
    <hr />
    <div
      class="justify2"
    >
      <div
        class="buttonNumero"
      >
        <button>
          -
        </button>
        <input
          class="CantidadTesting"
          type="text"
          value="0"
        />
        <button
          class="addCartTesting"
        >
          +
        </button>
      </div>
      <div
        class="rigth"
      >
        total $ 
        0
      </div>
    </div>
    <div
      class=""
    >
      <div>
        <button
          class="btn btn-2"
        >
          Add to Favorite
        </button>
        <button
          class="btn btn-2 black"
          disabled=""
        >
          Add to Cart
        </button>
      </div>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </p>
    </div>
  </div>
</div>
`;
